src/Microsoft.Health.Fhir.TemplateManagement/Overlay/OverlayFileSystem.cs

```diff
@@ -6,130 +6,155 @@
         {
-            return ReadOCIArtifactLayers(WorkingImageLayerFolder);
+            return await ReadOciArtifactLayersAsync(_workingImageLayerFolder);
         }
 
-        public void WriteImageLayers(List<OCIArtifactLayer> imageLayers)
+        public async Task WriteImageLayersAsync(List<ArtifactBlob> imageLayers)
         {
             EnsureArg.IsNotNull(imageLayers, nameof(imageLayers));
 
-            ClearFolder(WorkingImageLayerFolder);
+            Directory.CreateDirectory(_workingImageLayerFolder);
+
+            // Used to label layers sequence. In this version, only two layers will be generated.
+            // Should not exceed 9 in the future.
+            var layerNumber = 1;
             foreach (var layer in imageLayers)
             {
-                layer.WriteToFolder(WorkingImageLayerFolder);
+                if (layer.Content == null)
+                {
+                    continue;
+                }
+
+                await layer.WriteToFileAsync(Path.Combine(_workingImageLayerFolder, string.Format("layer{0}.tar.gz", layerNumber)));
+                layerNumber += 1;
             }
         }
 
-        public List<OCIArtifactLayer> ReadBaseLayers()
+        public async Task WriteManifestAsync(ManifestWrapper manifest)
         {
-            return ReadOCIArtifactLayers(WorkingBaseLayerFolder);
+            EnsureArg.IsNotNull(manifest, nameof(manifest));
+
+            Directory.CreateDirectory(_workingImageFolder);
+            await File.WriteAllTextAsync(Path.Combine(_workingImageFolder, Constants.ManifestFileName), JsonConvert.SerializeObject(manifest));
         }
 
-        public void WriteBaseLayers(List<OCIArtifactLayer> layers)
+        public async Task<ArtifactBlob> ReadBaseLayerAsync()
         {
-            EnsureArg.IsNotNull(layers, nameof(layers));
-
-            ClearFolder(WorkingBaseLayerFolder);
-            for (var index = 1; index <= layers.Count; index++)
+            var layers = await ReadOciArtifactLayersAsync(_workingBaseLayerFolder);
+            if (layers.Count() > 1)
             {
-                if (layers[index - 1].SequenceNumber == index)
-                {
-                    layers[index - 1].WriteToFolder(WorkingBaseLayerFolder);
-                }
+                throw new OverlayException(TemplateManagementErrorCode.BaseLayerLoadFailed, $"Base layer load failed. More than one layer in the base layer folder.");
             }
+
+            return layers.Count == 0 ? new OciFileLayer() : layers[0];
         }
 
-        public void ClearImageLayerFolder()
+        public async Task WriteBaseLayerAsync(ArtifactBlob baseLayer)
         {
-            ClearFolder(WorkingImageLayerFolder);
+            EnsureArg.IsNotNull(baseLayer, nameof(baseLayer));
+
+            Directory.CreateDirectory(_workingBaseLayerFolder);
+            await baseLayer.WriteToFileAsync(Path.Combine(_workingBaseLayerFolder, "layer1.tar.gz"));
```

Comment:
We can keep this into a constant for base layer name as well.

Reply:
Thanks, added into constants.